orte/tools/orte-restart/orte-restart.1
.\"
.\" Copyright (c) 2004-2007 The Trustees of Indiana University and Indiana
.\"                         University Research and Technology
.\"                         Corporation.  All rights reserved.
.\"
.\" Man page for OMPI's ompi-restart command
.\" 
.\" .TH name     section center-footer   left-footer  center-header
.TH     OMPI-RESTART  1       "March 2007" "Open MPI"   "OPEN MPI COMMANDS"
.\" **************************
.\"    Name Section
.\" **************************
.SH NAME
.
ompi-restart, orte-restart \- Restart a previously checkpointed parallel job
using the Open PAL Checkpoint/Restart Service (CRS)
.
.PP
.
\fBNOTE:\fP \fIompi-restart\fP, and \fIorte-restart\fP are all exact
synonyms for each other. Using any of the names will result in exactly
identical behavior.
.
.\" **************************
.\"    Synopsis Section
.\" **************************
.SH SYNOPSIS
.
.B ompi-restart
.R [ options ]
.B <GLOBAL SNAPSHOT HANDLE>
.
.\" **************************
.\"    Options Section
.\" **************************
.SH Options
.
\fIompi-restart\fR will attempt to restart a previously checkpointed parallel
job from the global snapshot handle reference returned by \fIompi_checkpoint\fP.
.
.TP 10
.B <GLOBAL SNAPSHOT HANDLE>
The global snapshot handle reference returned by \fIompi_checkpoint\fP, used to
restart the job. This is required to be the last argument to this command.
.
.
.TP
.B -h | --help
Display help for this command
.
.
.TP
.B -p | --preload
Preload the checkpoint files on the remote systems before restarting the
application. Disabled by default.
.
.
.TP
.B --fork
Fork off a new process, which is the restarted process. By default, the
restarted process will replace \fIompi-restart\fR.
.
.
.TP
.B -s | --seq
The sequence number of the checkpoint to restart from. By default, the most
recent sequence number is used (specified by -1).
.
.
.TP
.B -hostfile | --hostfile
The hostfile from which to restart the application. Useful in unscheduled
environments. (Same behavior as --machinefile option)
.
.
.TP
.B -machinefile | --machinefile
The machinefile from which to restart the application. Useful in unscheduled
environments. (Same behavior as --hostfile option)
.
.
.TP
.B -v | --verbose
Enable verbose output for debugging.
.
.
.TP
.B -gmca | --gmca \fR<key> <value>\fP
Pass global MCA parameters that are applicable to all contexts. \fI<key>\fP is
the parameter name; \fI<value>\fP is the parameter value.
.
.
.TP
.B -mca | --mca <key> <value>
Send arguments to various MCA modules.
.
.
.\" **************************
.\"    Description Section
.\" **************************
.SH DESCRIPTION
.
.PP
\fIompi-restart\fR can be invoked multiple, non-overlapping times. This
allows the user to restart a previously running parallel job.
.
.
.\" **************************
.\"    See Also Section
.\" **************************
.
.SH SEE ALSO
  orte-ps(1), orte-clean(1), ompi-checkpoint(1), opal-checkpoint(1), opal-restart(1), opal_crs(7)
.


NEWS
Copyright (c) 2004-2007 The Trustees of Indiana University and Indiana
                        University Research and Technology
                        Corporation.  All rights reserved.
Copyright (c) 2004-2006 The University of Tennessee and The University
                        of Tennessee Research Foundation.  All rights
                        reserved.
Copyright (c) 2004-2006 High Performance Computing Center Stuttgart, 
                        University of Stuttgart.  All rights reserved.
Copyright (c) 2004-2006 The Regents of the University of California.
                        All rights reserved.
Copyright (c) 2006-2007 Cisco Systems, Inc.  All rights reserved.
Copyright (c) 2006      Voltaire, Inc. All rights reserved.
Copyright (c) 2006      Sun Microsystems, Inc.  All rights reserved.
                        Use is subject to license terms.
Copyright (c) 2006-2007 Los Alamos National Security, LLC.  All rights
                        reserved.
$COPYRIGHT$

Additional copyrights may follow

$HEADER$

===========================================================================

This file contains the main features as well as overviews of specific
bug fixes (and other actions) for each version of Open MPI since
version 1.0.

1.2.1
-----

- Fixed issues with backtrace code on 64 bit Intel & PPC OS X builds.
- Fixed issues with multi-word CC variables and libtool.
  Thanks to Bert Wesarg for the bug reports.
- Fix issue with node naming schemes in SLURM
- Fix file descriptor leak in the Grid Engine/N1GE support
- Fix compile error on OS X 10.3.x introduced with Open MPI 1.1.5.
- Implement MPI_TYPE_CREATE_DARRAY function (was in 1.1.5 but not 1.2)
- Recognize zsh in rsh pls


1.2
---

- Fixed race condition in the shared memory fifo's, which led to
  orphaned messages.
- Corrected the size of the shared memory file - subtracted out the
  space the header was occupying.
- Add support for MPI_2COMPLEX and MPI_2DOUBLE_COMPLEX.
- Always ensure to create $(includedir)/openmpi, even if the C++
  bindings are disabled so that the wrapper compilers don't point to
  a directory that doesn't exist.  Thanks to Martin Audet for
  identifying the problem.
- Fixes for endian handling in MPI process startup.
- Openib BTL initialization fixes for cases where MPI processes in the
  same job has different numbers of active ports on the same physical
  fabric.
- Print more descriptive information when displaying backtraces on
  OS's that support this functionality, such as the hostname and PID
  of the process in question.
- Fixes to properly handle MPI exceptions in C++ on communicators,
  windows, and files.
- Much more reliable runtime support, particularly with regards to MPI
  job startup scalability, BProc support, and cleanup in failure
  scenarios (e.g., MPI_ABORT, MPI processes abnormally terminating,
  etc.).
- Significant performance improvements for MPI collectives,
  particularly on high-speed networks.
- Various fixes in the MX BTL component.
- Fix C++ typecast problems with MPI_ERRCODES_IGNORE.  Thanks to
  Satish Balay for bringing this to our attention.
- Allow run-time specification of the maximum amount of registered
  memory for OpenFabrics and GM.
- Users who utilize the wrapper compilers (e.g., mpicc and mpif77)
  will not notice, but the underlying library names for ORTE and OPAL
  have changed to libopen-rte and libopen-pal, respectively (listed
  here because there are undoubtedly some users who are not using the
  wrapper compilers).
- Many bug fixes to MPI-2 one-sided support.
- Added support for TotalView message queue debugging.
- Fixes for MPI_STATUS_SET_ELEMENTS.
- Print better error messages when mpirun's "-nolocal" is used when
  there is only one node available.
- Added man pages for several Open MPI executables and the MPI API
  functions.
- A number of fixes for Alpha platforms.
- A variety of Fortran API fixes.
- Build the Fortran MPI API as a separate library to allow these
  functions to be profiled properly.
- Add new --enable-mpirun-prefix-by-default configure option to always
  imply the --prefix option to mpirun, preventing many rsh/ssh-based
  users from needing to modify their shell startup files.
- Add a number of missing constants in the C++ bindings.
- Added tight integration with Sun N1 Grid Engine (N1GE) 6 and the
  open source Grid Engine.
- Allow building the F90 MPI bindings as shared libraries for most
  compilers / platforms.  Explicitly disallow building the F90
  bindings as shared libraries on OS X because of complicated
  situations with Fortran common blocks and lack of support for
  unresolved common symbols in shared libraries.
- Added stacktrace support for Solaris and Mac OS X.
- Update event library to libevent-1.1b.
- Fixed standards conformance issues with MPI_ERR_TRUNCATED and
  setting MPI_ERROR during MPI_TEST/MPI_WAIT.
- Addition of "cm" PML to better support library-level matching
  interconnects, with support for Myrinet/MX, and QLogic PSM-based
  networks.
- Addition of "udapl" BTL for transport across uDAPL interconnects.
- Really check that the $CXX given to configure is a C++ compiler
  (not a C compiler that "sorta works" as a C++ compiler).
- Properly check for local host only addresses properly, looking 
  for 127.0.0.0/8, rather than just 127.0.0.1.


1.1.5
-----

- Fix syntax error in a corner case of the event library.  Thanks to
  Bert Wesarg for pointing this out.
- Add new MCA parameter (mpi_preconnect_oob) for pre-connecting the
  "out of band" channels between all MPI processes.  Most helpful for
  MPI applications over InfiniBand where process A sends an initial
  message to process B, but process B does not enter the MPI library
  for a long time.
- Fix for a race condition in shared memory locking semantics.
- Add major, minor, and release version number of Open MPI to mpi.h.
  Thanks to Martin Audet for the suggestion.
- Fix the "restrict" compiler check in configure.

orte/tools/orte-restart/orte-restart.c
/*
 * Copyright (c) 2004-2007 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2005 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart, 
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * $COPYRIGHT$
 * 
 * Additional copyrights may follow
 * 
 * $HEADER$
 */

/**
 * @file
 * ORTE Restart Tool for restarting a previously checkpointed multiprocess job
 *
 */

#include "orte_config.h"

#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <libgen.h>
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif  /*  HAVE_STDLIB_H */
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#ifdef HAVE_LIBGEN_H
#include <libgen.h>
#endif
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#endif  /* HAVE_STRING_H */

#include "opal/runtime/opal.h"
#include "opal/runtime/opal_cr.h"
#include "opal/util/cmd_line.h"
#include "opal/util/argv.h"
#include "opal/util/show_help.h"
#include "opal/util/output.h"
#include "opal/util/opal_environ.h"
#include "opal/util/os_path.h"
#include "opal/mca/base/base.h"
#include "opal/mca/base/mca_base_param.h"
#include "opal/mca/crs/crs.h"
#include "opal/mca/crs/base/base.h"

#include "orte/orte_constants.h"
#include "orte/runtime/runtime.h"
#include "orte/runtime/orte_cr.h"
#include "orte/mca/snapc/snapc.h"
#include "orte/mca/snapc/base/base.h"
#include "orte/mca/filem/filem.h"
#include "orte/mca/filem/base/base.h"

/*****************
 * Global Vars
 *****************/
extern char **environ;

/******************
 * Local Functions
 ******************/
static int initialize(int argc, char *argv[]);
static int finalize(void);
static int parse_args(int argc, char *argv[]);
static int check_file(orte_snapc_base_global_snapshot_t *snapshot);
static int create_appfile(orte_snapc_base_global_snapshot_t *snapshot);
static int spawn_children(orte_snapc_base_global_snapshot_t *snapshot, pid_t *child_pid);

/*****************************************
 * Global Vars for Command line Arguments
 *****************************************/
typedef struct {
    bool help;
    char *filename;
    char *appfile;
    bool verbose;
    bool forked;
    bool preload;
    int  seq_number;
    char *hostfile;
    int  output;
} orte_restart_globals_t;

orte_restart_globals_t orte_restart_globals;

opal_cmd_line_init_t cmd_line_opts[] = {
    { NULL, NULL, NULL, 
      'h', NULL, "help", 
      0,
      &orte_restart_globals.help, OPAL_CMD_LINE_TYPE_BOOL,
      "This help message" },

    { NULL, NULL, NULL, 
      'v', NULL, "verbose", 
      0,
      &orte_restart_globals.verbose, OPAL_CMD_LINE_TYPE_BOOL,
      "Be Verbose" },

    { NULL, NULL, NULL, 
      'p', NULL, "preload", 
      0,
      &orte_restart_globals.preload, OPAL_CMD_LINE_TYPE_BOOL,
      "Preload the checkpoint files before restarting (Default = Disabled)" },

    { NULL, NULL, NULL, 
      '\0', NULL, "fork", 
      0,
      &orte_restart_globals.forked, OPAL_CMD_LINE_TYPE_BOOL,
      "Fork off a new process which is the restarted process instead of "
      "replacing orte_restart" },

    { NULL, NULL, NULL, 
      's', NULL, "seq", 
      1,
      &orte_restart_globals.seq_number, OPAL_CMD_LINE_TYPE_INT,
      "The sequence number of the checkpoint to start from. "
      "(Default: -1, or most recent)" },

    { NULL, NULL, NULL, 
      '\0', "hostfile", "hostfile", 
      1,
      &orte_restart_globals.hostfile, OPAL_CMD_LINE_TYPE_STRING,
      "Provide a hostfile to use for launch" },

    { NULL, NULL, NULL, 
      '\0', "machinefile", "machinefile", 
      1,
      &orte_restart_globals.hostfile, OPAL_CMD_LINE_TYPE_STRING,
      "Provide a hostfile to use for launch" },

    /* End of list */
    { NULL, NULL, NULL, 
      '\0', NULL, NULL, 
      0,
      NULL, OPAL_CMD_LINE_TYPE_NULL,
      NULL }
};

int
main(int argc, char *argv[])
{
    int ret, exit_status = ORTE_SUCCESS;
    pid_t child_pid;
    orte_snapc_base_global_snapshot_t *snapshot = NULL;
    
    /***************
     * Initialize
     ***************/
    if (ORTE_SUCCESS != (ret = initialize(argc, argv))) {
        exit_status = ret;
        goto cleanup;
    }

    snapshot = OBJ_NEW(orte_snapc_base_global_snapshot_t);
    snapshot->reference_name  = strdup(orte_restart_globals.filename);
    snapshot->local_location  = dirname(orte_snapc_base_get_global_snapshot_directory(snapshot->reference_name));

    /* 
     * Check for existence of the file
     */
    if( ORTE_SUCCESS != (ret = check_file(snapshot)) ) {
        opal_show_help("help-orte-restart.txt", "invalid_filename", true,
                       orte_restart_globals.filename);
        exit_status = ret;
        goto cleanup;
    }

    /******************************
     * Create the app file to use with orterun
     ******************************/
    if( ORTE_SUCCESS != (ret = create_appfile(snapshot) ) ) {
        exit_status = ret;
        goto cleanup;
    }

    /******************************
     * Restart in this process [orterun]
     ******************************/
    if( orte_restart_globals.verbose ) {
        opal_output_verbose(10, orte_restart_globals.output,
                            "Restarting from file (%s)",
                            orte_restart_globals.filename);
        
        if( orte_restart_globals.forked ) {
            opal_output_verbose(10, orte_restart_globals.output,
                                "\t Forking off a child");
        } else {
            opal_output_verbose(10, orte_restart_globals.output,
                                "\t Exec in self");
        }
    }

    if( ORTE_SUCCESS != (ret = spawn_children(snapshot, &child_pid)) ) {
        opal_show_help("help-orte-restart.txt", "restart_cmd_failure", true,
                       orte_restart_globals.filename, ret);
        exit_status = ret;
        goto cleanup;
    }

    opal_output_verbose(10, orte_restart_globals.output,
                        "orte_restart: Restarted Child with PID = %d\n", child_pid);

    /***************
     * Cleanup
     ***************/
 cleanup:
    if(NULL != snapshot )
        OBJ_RELEASE(snapshot);

    if (OPAL_SUCCESS != (ret = finalize())) {
        return ret;
    }

    return exit_status;
}

static int initialize(int argc, char *argv[]) {
    int ret, exit_status = ORTE_SUCCESS;

    /*
     * Parse command line arguments
     */
    if (ORTE_SUCCESS != (ret = parse_args(argc, argv))) {
        exit_status = ret;
        goto cleanup;
    }

    /*
     * Setup OPAL Output handle from the verbose argument
     */
    if( orte_restart_globals.verbose ) {
        orte_restart_globals.output = opal_output_open(NULL);
        opal_output_set_verbosity(orte_restart_globals.output, 10);
    } else {
        orte_restart_globals.output = 0; /* Default=STDOUT */
    }

    /* We don't need any crs component */
    opal_setenv(mca_base_param_env_var("crs"),
                strdup("none"),
                true, &environ);
    opal_cr_set_enabled(false);

    if (OPAL_SUCCESS != (ret = orte_init(true))) {
        exit_status = ret;
        goto cleanup;
    }
    
    /* Unset these now that we no longer need them */
    opal_unsetenv(mca_base_param_env_var("crs"), &environ);
    opal_unsetenv(mca_base_param_env_var("opal_cr_is_tool"), &environ);

 cleanup:
    return exit_status;
}

static int finalize(void) {
    int ret;

    if (OPAL_SUCCESS != (ret = orte_finalize())) {
        return ret;
    }

    return ORTE_SUCCESS;
}

static int parse_args(int argc, char *argv[]) {
    int i, ret, len;
    opal_cmd_line_t cmd_line;
    char **app_env = NULL, **global_env = NULL;
    orte_restart_globals_t tmp = { false, /* help */
                                   NULL,  /* filename */
                                   NULL,  /* appfile */
                                   false, /* verbose */
                                   false, /* forked */
                                   false, /* preload */
                                   -1,    /* seq_number */
                                   NULL,  /* hostfile */
                                   -1 };  /* output*/

    orte_restart_globals = tmp;

    /* Parse the command line options */    
    opal_cmd_line_create(&cmd_line, cmd_line_opts);
    
    mca_base_open();
    mca_base_cmd_line_setup(&cmd_line);
    ret = opal_cmd_line_parse(&cmd_line, true, argc, argv);
    
    /** 
     * Put all of the MCA arguments in the environment 
     */
    mca_base_cmd_line_process_args(&cmd_line, &app_env, &global_env);
    
    len = opal_argv_count(app_env);
    for(i = 0; i < len; ++i) {
        putenv(app_env[i]);
    }

    len = opal_argv_count(global_env);
    for(i = 0; i < len; ++i) {
        putenv(global_env[i]);
    }
    

    opal_setenv(mca_base_param_env_var("opal_cr_is_tool"),
                "1",
                true, &environ);
    /**
     * Now start parsing our specific arguments
     */
    if (OPAL_SUCCESS != ret || 
        orte_restart_globals.help ||
        1 >= argc) {
        char *args = NULL;
        args = opal_cmd_line_get_usage_msg(&cmd_line);
        opal_show_help("help-orte-restart.txt", "usage", true,
                       args);
        free(args);
        return ORTE_ERROR;
    }

    /* get the remaining bits */
    opal_cmd_line_get_tail(&cmd_line, &argc, &argv);
    if ( 1 > argc ) {
        char *args = NULL;
        args = opal_cmd_line_get_usage_msg(&cmd_line);
        opal_show_help("help-orte-restart.txt", "usage", true,
                       args);
        free(args);
        return ORTE_ERROR;
    }

    orte_restart_globals.filename = strdup(argv[0]);
    if ( NULL == orte_restart_globals.filename || 
         0 >= strlen(orte_restart_globals.filename) ) {
        opal_show_help("help-orte-restart.txt", "invalid_filename", true,
                       orte_restart_globals.filename);
        return ORTE_ERROR;
    }

    /* If we have arguments after the command, then assume they
     * need to be grouped together.
     */
    if(argc > 1) {
        orte_restart_globals.filename = strdup(opal_argv_join(argv, ' '));
    }
    
    return ORTE_SUCCESS;
}

static int check_file(orte_snapc_base_global_snapshot_t *snapshot) {
    int ret, exit_status = ORTE_SUCCESS;

    opal_output_verbose(10, orte_restart_globals.output,
                        "Checking for the existence of (%s)\n",
                        snapshot->local_location);

    if (0 >  (ret = access(snapshot->local_location, F_OK)) ) {
        exit_status = ORTE_ERROR;
        goto cleanup;
    }

 cleanup:
    return exit_status;
}

static int create_appfile(orte_snapc_base_global_snapshot_t *snapshot) {
    int ret, exit_status = ORTE_SUCCESS;
    FILE *appfile = NULL;
    opal_list_item_t* item = NULL;

    /*
     * Extract the record information for the specified seq number.
     * Note: If the seq # passed is -1, then the largest seq # is selected,
     *       ow the seq # requested is selected if available
     */
    snapshot->seq_num = orte_restart_globals.seq_number;
    if( ORTE_SUCCESS != (ret = orte_snapc_base_extract_metadata( snapshot ) ) ) {
        exit_status = ret;
        goto cleanup;
    }

    /*
     * Create the appfile
     */
    asprintf(&orte_restart_globals.appfile, "%s/%s",
             snapshot->local_location,
             strdup("restart-appfile"));

    if (NULL == (appfile = fopen(orte_restart_globals.appfile, "w")) ) {
        exit_status = ret;
        goto cleanup;
    }

    for(item  = opal_list_get_first(&snapshot->snapshots);
        item != opal_list_get_end(&snapshot->snapshots);
        item  = opal_list_get_next(item) ) {
        orte_snapc_base_snapshot_t *vpid_snapshot;
        vpid_snapshot = (orte_snapc_base_snapshot_t*)item;
        
        fprintf(appfile, "#\n");
        fprintf(appfile, "# Old Process Name: %u.%u.%u\n", 
                vpid_snapshot->process_name.cellid,
                vpid_snapshot->process_name.jobid,
                vpid_snapshot->process_name.vpid);
        fprintf(appfile, "#\n");
        fprintf(appfile, "-np 1 ");
        if(orte_restart_globals.preload) {
            fprintf(appfile, "--preload-files %s/%s ", 
                    vpid_snapshot->crs_snapshot_super.local_location, 
                    vpid_snapshot->crs_snapshot_super.reference_name);
            fprintf(appfile, "--preload-files-dest-dir . ");
        }
        /* JJH: Make this match what the user originally specified on the command line */
        fprintf(appfile, "-am ft-enable-cr ");
        fprintf(appfile, " opal-restart ");
        /* JJH: Make sure this changes if ever the default location of the local file is changed,
         * currently it is safe to assume that it is in the current working directory.
         *
         * JJH: If we allow inplace restarting then this may be another directory... */
        if(orte_restart_globals.preload) {
            /* If we preloaded the files then they are in the current working
             * directory. */
            fprintf(appfile, "-mca crs_base_snapshot_dir . ");
        }
        else {
            /* If we are *not* preloading the files, the point to the original checkpoint
             * directory to access the checkpoint files. */
            fprintf(appfile, "-mca crs_base_snapshot_dir %s ", vpid_snapshot->crs_snapshot_super.local_location);
        }
        fprintf(appfile, "%s\n", vpid_snapshot->crs_snapshot_super.reference_name);
    }

 cleanup:
    if(NULL != appfile)
        fclose(appfile);
    
    return exit_status;
}

static int spawn_children(orte_snapc_base_global_snapshot_t *snapshot, pid_t *child_pid) {
    int ret, exit_status = ORTE_SUCCESS;
    char **argv = NULL;
    int argc = 0;
    int status;

    if( ORTE_SUCCESS != (ret = opal_argv_append(&argc, &argv, "orterun")) ) {
        exit_status = ret;
        goto cleanup;
    }
    if( NULL != orte_restart_globals.hostfile ) {
        if( ORTE_SUCCESS != (ret = opal_argv_append(&argc, &argv, "--hostfile")) ) {
            exit_status = ret;
            goto cleanup;
        }
        if( ORTE_SUCCESS != (ret = opal_argv_append(&argc, &argv, orte_restart_globals.hostfile)) ) {
            exit_status = ret;
            goto cleanup;
        }
    }
    if( ORTE_SUCCESS != (ret = opal_argv_append(&argc, &argv, "--app")) ) {
        exit_status = ret;
        goto cleanup;
    }
    if( ORTE_SUCCESS != (ret = opal_argv_append(&argc, &argv, orte_restart_globals.appfile)) ) {
        exit_status = ret;
        goto cleanup;
    }

    /* To fork off a child */
    if( orte_restart_globals.forked ) {
        *child_pid = fork();
        
        if( 0 == *child_pid) {
            /* Child Process */
            status = execvp(strdup(argv[0]), argv);
            if( 0 > status) {
                opal_output(orte_restart_globals.output,
                            "orte_restart: execv failed with status = %d\n",
                            status);
            }
            exit_status = status;
            goto cleanup;
        }
        else if(0 < *child_pid) {
            /* Parent is done once it is started */
            ;
        }
        else {
            opal_output(orte_restart_globals.output,
                        "orte_restart: fork failed: This should never happen!");
            /* Fork failed :( */
            exit_status = *child_pid;
            goto cleanup;
        }
    }
    /* ... or not to fork off a child */
    else {
        /* Make sure to finalize so we don't leave our session directory */
        orte_system_finalize();
        /* reopen the crs components */
        opal_crs_base_open();

        status = execvp(strdup(argv[0]), argv);
        if( 0 > status) {
            /* execv failed */
        }
        exit_status = status;
        goto cleanup;
    }

 cleanup:
    if( NULL != argv)
        opal_argv_free(argv);

    return exit_status;
}
